2006-01-08  Ian Lance Taylor  <ian@airs.com>
            David Edelsohn  <edelsohn@gnu.org>

            PR rtl-optimization/25662
              * optabs.c (simplify_expand_binop): Use
              * simplify_binary_operation for constant operands instead of
                simplify_gen_binary.

--- gcc/optabs.c	2006/01/08 20:54:28	109476
+++ gcc/optabs.c	2006/01/08 20:55:39	109477
@@ -416,9 +416,14 @@
 		       enum optab_methods methods)
 {
   if (CONSTANT_P (op0) && CONSTANT_P (op1))
-    return simplify_gen_binary (binoptab->code, mode, op0, op1);
-  else
-    return expand_binop (mode, binoptab, op0, op1, target, unsignedp, methods);
+    {
+      rtx x = simplify_binary_operation (binoptab->code, mode, op0, op1);
+
+      if (x)
+	return x;
+    }
+
+  return expand_binop (mode, binoptab, op0, op1, target, unsignedp, methods);
 }
 
 /* Like simplify_expand_binop, but always put the result in TARGET.
